File: ios/Src/Mytiki/Sdk/Capture/Receipt/Capacitor/Rsp/RspProduct.swift
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import BlinkReceipt
import BlinkEReceipt
import Capacitor

struct RspProduct : Rsp {
    private let productNumber: RspStringType?
    private let description: RspStringType?
    private let quantity: RspFloatType?
    private let unitPrice: RspFloatType?
    private let unitOfMeasure: RspStringType?
    private let totalPrice: RspFloatType?
    private let fullPrice: Float
    private let line: Int
    private let productName: String?
    private let brand: String?
    private let category: String?
    private let size: String?
    private let rewardsGroup: String?
    private let competitorRewardsGroup: String?
    private let upc: String?
    private let imageUrl: String?
    private let shippingStatus: String?
    private let additionalLines: List<RspAdditionalLine>
    private let priceAfterCoupons: RspFloatType?
    private let voided: Bool
    private let probability: Double
    private let sensitive: Bool
    private let possibleProducts: List<RspProduct>
    private let subProducts: List<RspProduct>
    private let added: Bool
    private let blinkReceiptBrand: String?
    private let blinkReceiptCategory: String?
    private let extendedFields: JSONObject?
    private let fuelType: String?
    private let descriptionPrefix: RspStringType?
    private let descriptionPostfix: RspStringType?
    private let skuPrefix: RspStringType?
    private let skuPostfix: RspStringType?
    private let attributes: List<JSObject>
    private let sector: String?
    private let department: String?
    private let majorCategory: String?
    private let subCategory: String?
    private let itemType: String?
    
    init (product: Product){
        productNumber = RspStringType.opt(product.productNumber())
        self.description = RspStringType.opt(product.description())
        self.quantity = RspFloatType.opt(product.quantity())
        self.unitPrice = RspFloatType.opt(product.unitPrice())
        self.unitOfMeasure = RspStringType.opt(product.unitOfMeasure())
        self.totalPrice = RspFloatType.opt(product.totalPrice())
        fullPrice = product.fullPrice()
        line = product.line()
        productName = product.productName()
        brand = product.brand()
        category = product.category()
        size = product.size()
        rewardsGroup = product.rewardsGroup()
        competitorRewardsGroup = product.competitorRewardsGroup()
        upc = product.upc()
        imageUrl = product.imageUrl()
        shippingStatus = product.shippingStatus()
        additionalLines =
            product.additionalLines()?.map { additionalLine -> RspAdditionalLine(additionalLine) }
                ?? emptyList()
        priceAfterCoupons = RspFloatType.opt(product.priceAfterCoupons())
        voided = product.voided()
        probability = product.probability()
        sensitive = product.sensitive()
        possibleProducts = product.possibleProducts()?.map { prd -> RspProduct(prd) } ?? emptyList()
        subProducts = product.subProducts()?.map { prd -> RspProduct(prd) } ?? emptyList()
        added = product.added()
        blinkReceiptBrand = product.blinkReceiptBrand()
        blinkReceiptCategory = product.blinkReceiptCategory()
        fuelType = product.fuelType()
        descriptionPrefix = RspStringType.opt(product.descriptionPrefix())
        descriptionPostfix = RspStringType.opt(product.descriptionPostfix())
        skuPrefix = RspStringType.opt(product.skuPrefix())
        skuPostfix = RspStringType.opt(product.skuPostfix())
        sector = product.sector()
        department = product.department()
        majorCategory = product.majorCategory()
        subCategory = product.subCategory()
        itemType = product.itemType()
        extendedFields =  (product.extendedFields() != null) ? {
            val; extendedFields = JSObject()
            product.extendedFields()
                .forEach { entry -> extendedFields.put(entry.key, entry.value) }
            extendedFields
        } : nil
        attributes = (product.attributes() != null) ? {
            product.attributes()!!.map { attr ->
                val; json = JSONObject()
                attr.forEach { entry -> json.put(entry.key, entry.value) }
                json
            }
        } : emptyList()
    }
    
    func toJson() -> JSObject {
        JSObject.updateValue("productNumber", productNumber?.toJson())
        JSObject.updateValue("description", description?.toJson())
        JSObject.updateValue("quantity", quantity?.toJson())
        JSObject.updateValue("unitPrice", unitPrice?.toJson())
        JSObject.updateValue("unitOfMeasure", unitOfMeasure?.toJson())
        JSObject.updateValue("totalPrice", totalPrice?.toJson())
        JSObject.updateValue("fullPrice", fullPrice)
        JSObject.updateValue("line", line)
        JSObject.updateValue("productName", productName)
        JSObject.updateValue("brand", brand)
        JSObject.updateValue("category", category)
        JSObject.updateValue("size", size)
        JSObject.updateValue("rewardsGroup", rewardsGroup)
        JSObject.updateValue("competitorRewardsGroup", competitorRewardsGroup)
        JSObject.updateValue("upc", upc)
        JSObject.updateValue("imageUrl", imageUrl)
        JSObject.updateValue("shippingStatus", shippingStatus)
        JSObject.updateValue("additionalLines", JSONArray(additionalLines.map { lineJSObject.updateValueline.toJson() }))
        JSObject.updateValue("priceAfterCoupons", priceAfterCoupons?.toJson())
        JSObject.updateValue("voided", voided)
        JSObject.updateValue("probability", probability)
        JSObject.updateValue("sensitive", sensitive)
        JSObject.updateValue("possibleProducts", JSONArray(possibleProducts.map { prdJSObject.updateValueprd.toJson() }))
        JSObject.updateValue("subProducts", JSONArray(subProducts.map { prd -> prd.toJson() }))
        JSObject.updateValue("added", added)
        JSObject.updateValue("blinkReceiptBrand", blinkReceiptBrand)
        JSObject.updateValue("blinkReceiptCategory", blinkReceiptCategory)
        JSObject.updateValue("extendedFields", extendedFields)
        JSObject.updateValue("fuelType", fuelType)
        JSObject.updateValue("descriptionPrefix", descriptionPrefix?.toJson())
        JSObject.updateValue("descriptionPostfix", descriptionPostfix?.toJson())
        JSObject.updateValue("skuPrefix", skuPrefix?.toJson())
        JSObject.updateValue("skuPostfix", skuPostfix?.toJson())
        JSObject.updateValue("attributes", JSONArray(attributes))
        JSObject.updateValue("sector", sector)
        JSObject.updateValue("department", department)
        JSObject.updateValue("majorCategory", majorCategory)
        JSObject.updateValue("subCategory", subCategory)
        JSObject.updateValue("itemType", itemType)
    }
}

File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/rsp/RspProduct.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.rsp

import com.microblink.core.Product
import org.json.JSONArray
import org.json.JSONObject

/**
 * Represents a parsed product from a receipt using the RSP (Receipt Structure Parser) framework.
 *
 * @property productNumber The product number, if available.
 * @property description The product description, if available.
 * @property quantity The quantity of the product, if available.
 * @property unitPrice The unit price of the product, if available.
 * @property unitOfMeasure The unit of measure for the product, if available.
 * @property totalPrice The total price of the product, if available.
 * @property fullPrice The full price of the product.
 * @property line The line number of the product on the receipt.
 * @property productName The product name, if available.
 * @property brand The product brand, if available.
 * @property category The product category, if available.
 * @property size The product size, if available.
 * @property rewardsGroup The rewards group associated with the product, if available.
 * @property competitorRewardsGroup The competitor's rewards group, if available.
 * @property upc The Universal Product Code (UPC) of the product, if available.
 * @property imageUrl The URL of the product image, if available.
 * @property shippingStatus The shipping status of the product, if available.
 * @property additionalLines Additional lines of text associated with the product.
 * @property priceAfterCoupons The product price after applying coupons, if available.
 * @property voided Indicates whether the product is voided.
 * @property probability The probability score associated with the product.
 * @property sensitive Indicates whether the product is sensitive.
 * @property possibleProducts List of possible sub-products associated with the product.
 * @property subProducts List of sub-products associated with the product.
 * @property added Indicates whether the product was added.
 * @property blinkReceiptBrand The brand information from the BlinkReceipt SDK, if available.
 * @property blinkReceiptCategory The category information from the BlinkReceipt SDK, if available.
 * @property extendedFields Additional extended fields associated with the product.
 * @property fuelType The fuel type of the product, if available.
 * @property descriptionPrefix The prefix of the product description, if available.
 * @property descriptionPostfix The postfix of the product description, if available.
 * @property skuPrefix The prefix of the product SKU, if available.
 * @property skuPostfix The postfix of the product SKU, if available.
 * @property attributes List of attributes associated with the product.
 * @property sector The sector of the product, if available.
 * @property department The department of the product, if available.
 * @property majorCategory The major category of the product, if available.
 * @property subCategory The sub-category of the product, if available.
 * @property itemType The item type of the product, if available.
 */
class RspProduct(product: Product) : Rsp {
    private val productNumber: RspStringType?
    private val description: RspStringType?
    private val quantity: RspFloatType?
    private val unitPrice: RspFloatType?
    private val unitOfMeasure: RspStringType?
    private val totalPrice: RspFloatType?
    private val fullPrice: Float
    private val line: Int
    private val productName: String?
    private val brand: String?
    private val category: String?
    private val size: String?
    private val rewardsGroup: String?
    private val competitorRewardsGroup: String?
    private val upc: String?
    private val imageUrl: String?
    private val shippingStatus: String?
    private val additionalLines: List<RspAdditionalLine>
    private val priceAfterCoupons: RspFloatType?
    private val voided: Boolean
    private val probability: Double
    private val sensitive: Boolean
    private val possibleProducts: List<RspProduct>
    private val subProducts: List<RspProduct>
    private val added: Boolean
    private val blinkReceiptBrand: String?
    private val blinkReceiptCategory: String?
    private val extendedFields: JSONObject?
    private val fuelType: String?
    private val descriptionPrefix: RspStringType?
    private val descriptionPostfix: RspStringType?
    private val skuPrefix: RspStringType?
    private val skuPostfix: RspStringType?
    private val attributes: List<JSONObject>
    private val sector: String?
    private val department: String?
    private val majorCategory: String?
    private val subCategory: String?
    private val itemType: String?

    init {
        productNumber = RspStringType.opt(product.productNumber())
        this.description = RspStringType.opt(product.description())
        this.quantity = RspFloatType.opt(product.quantity())
        this.unitPrice = RspFloatType.opt(product.unitPrice())
        this.unitOfMeasure = RspStringType.opt(product.unitOfMeasure())
        this.totalPrice = RspFloatType.opt(product.totalPrice())
        fullPrice = product.fullPrice()
        line = product.line()
        productName = product.productName()
        brand = product.brand()
        category = product.category()
        size = product.size()
        rewardsGroup = product.rewardsGroup()
        competitorRewardsGroup = product.competitorRewardsGroup()
        upc = product.upc()
        imageUrl = product.imageUrl()
        shippingStatus = product.shippingStatus()
        additionalLines =
            product.additionalLines()?.map { additionalLine -> RspAdditionalLine(additionalLine) }
                ?: emptyList()
        priceAfterCoupons = RspFloatType.opt(product.priceAfterCoupons())
        voided = product.voided()
        probability = product.probability()
        sensitive = product.sensitive()
        possibleProducts = product.possibleProducts()?.map { prd -> RspProduct(prd) } ?: emptyList()
        subProducts = product.subProducts()?.map { prd -> RspProduct(prd) } ?: emptyList()
        added = product.added()
        blinkReceiptBrand = product.blinkReceiptBrand()
        blinkReceiptCategory = product.blinkReceiptCategory()
        fuelType = product.fuelType()
        descriptionPrefix = RspStringType.opt(product.descriptionPrefix())
        descriptionPostfix = RspStringType.opt(product.skuPostfix())
        skuPrefix = RspStringType.opt(product.skuPrefix())
        skuPostfix = RspStringType.opt(product.skuPostfix())
        sector = product.sector()
        department = product.department()
        majorCategory = product.majorCategory()
        subCategory = product.subCategory()
        itemType = product.itemType()
        extendedFields = if (product.extendedFields() != null) {
            val extendedFields = JSONObject()
            product.extendedFields()
                ?.forEach { entry -> extendedFields.put(entry.key, entry.value) }
            extendedFields
        } else null
        attributes = if (product.attributes() != null) {
            product.attributes()!!.map { attr ->
                val json = JSONObject()
                attr.forEach { entry -> json.put(entry.key, entry.value) }
                json
            }
        } else emptyList()
    }

    /**
     * Converts the RSP product to a JSON representation.
     *
     * @return A JSON object representing the RSP product.
     */
    override fun toJson(): JSONObject =
        JSONObject()
            .put("productNumber", productNumber?.toJson())
            .put("description", description?.toJson())
            .put("quantity", quantity?.toJson())
            .put("unitPrice", unitPrice?.toJson())
            .put("unitOfMeasure", unitOfMeasure?.toJson())
            .put("totalPrice", totalPrice?.toJson())
            .put("fullPrice", fullPrice)
            .put("line", line)
            .put("productName", productName)
            .put("brand", brand)
            .put("category", category)
            .put("size", size)
            .put("rewardsGroup", rewardsGroup)
            .put("competitorRewardsGroup", competitorRewardsGroup)
            .put("upc", upc)
            .put("imageUrl", imageUrl)
            .put("shippingStatus", shippingStatus)
            .put("additionalLines", JSONArray(additionalLines.map { line -> line.toJson() }))
            .put("priceAfterCoupons", priceAfterCoupons?.toJson())
            .put("voided", voided)
            .put("probability", probability)
            .put("sensitive", sensitive)
            .put("possibleProducts", JSONArray(possibleProducts.map { prd -> prd.toJson() }))
            .put("subProducts", JSONArray(subProducts.map { prd -> prd.toJson() }))
            .put("added", added)
            .put("blinkReceiptBrand", blinkReceiptBrand)
            .put("blinkReceiptCategory", blinkReceiptCategory)
            .put("extendedFields", extendedFields)
            .put("fuelType", fuelType)
            .put("descriptionPrefix", descriptionPrefix?.toJson())
            .put("descriptionPostfix", descriptionPostfix?.toJson())
            .put("skuPrefix", skuPrefix?.toJson())
            .put("skuPostfix", skuPostfix?.toJson())
            .put("attributes", JSONArray(attributes))
            .put("sector", sector)
            .put("department", department)
            .put("majorCategory", majorCategory)
            .put("subCategory", subCategory)
            .put("itemType", itemType)

    companion object {
        /**
         * Creates an optional [RspProduct] from a [Product] instance.
         *
         * @param product The [Product] instance to create an [RspProduct] from.
         * @return An [RspProduct] instance or null if the input is null.
         */
        fun opt(product: Product?): RspProduct? =
            if (product != null) RspProduct(product) else null
    }
}
